### DISCUSSION TITLE
Implement MariaDB 10.11 Grammar

### DISCUSSION THREAD

**Discussion by juanparati** [laravel/framework#48888] [Ideas]
MariaDB is a popular RDBMS that is traditionally compatible with MySQL. However, since version 10.1 the MariaDB team started to implement some features that may differ or are not present in modern versions of MySQL. Currently, Laravel bundles a generic MySQL schema grammar (https://github.com/laravel/framework/tree/10.x/src/Illuminate/Database/Schema/Grammars) that limits the usage of MariaDB.

- MariaDB implemented a native UUID in v10.7 that may replace the usage of "char(36)" as UUID (https://mariadb.com/kb/en/uuid/)
- MariaDB supports native INET4 and INET6 fields that are internally compatible with binary representations of the IPv4 and IPv6.(https://mariadb.com/kb/en/inet6/)
- MariaDB integrated some engines not present in MySQL, such MyRocks, ColumnStorage, and Aria, that may implement additional features for field types.

MariaDB is an RDBMS that is more relevant in the industry and is a fair competitor of MySQL, and I believe that with time, the differences between both will be more significant.

**Comment by Jubeki** [comment]
This may be something for Laravel 11, where MariaDB gets their own config. (see https://github.com/laravel/framework/pull/48455)

It could be extended with a new MariaDB Driver. (In the beginning it can be a copy of the MySQL Grammar and then new methods can be added, or features can be changed which are specific for MariaDB)

You could try a PR, since Laravel 10 already supports MariaDB 10.10+ (see https://laravel.com/docs/10.x/database#introduction)

**Reply by juanparati** [reply]
@Jubeki : Thank you tell about. I will take a look  👀

**Reply by shaedrich** [reply]
Yeah, it would make sense to have a different driver then as well:
https://github.com/laravel/framework/blob/220e72240161cf1834cb268e68d1722e9dcc96d8/config/database.php#L62

**Reply by azerum** [reply]
Does 10.10+ mean '10.10.X' (any version with same minor), or any version starting from 10.10.0?

**Reply by Jubeki** [reply]
10.10+ stands for >= 10.10 so 10.10.x, 10.11.x, ..., 11.x

**Comment by isra-skyler** [comment]
any version of MariaDB that is 10.10 or higher. I just got it working on a recent project of mine the client was looking to [Hire Laravel Developers in Marysville](https://hybridwebagency.com/marysville-wa/hire-laravel-developers/)

**Comment by staudenmeir** [comment]
I've created a PR for a MariaDB driver: #50146

**Comment by staudenmeir** [comment]
The MariaDB driver will be included in Laravel 11.

> MariaDB implemented a native UUID in v10.7 that may replace the usage of "char(36)" as UUID (https://mariadb.com/kb/en/uuid/)

MariaDB migrations will use the native UUID type (#50192, https://github.com/laravel/framework/commit/8648a4a8ff761b01d9a390b9ddb4cdd46201e86d). When you switch to the MariaDB driver, you need to convert all existing UUID columns to the new type to prevent breaking changes:

```php
Schema::table('your_table', function (Blueprint $table) {
    $table->uuid('your_uuid')->change();
});
```

> MariaDB supports native INET4 and INET6 fields that are internally compatible with binary representations of the IPv4 and IPv6.(https://mariadb.com/kb/en/inet6/)

AFAIU, we can't benefit from these since Laravel only has a general column type `ipAddress()` for both IPv4 and IPv6 addresses. The MariaDB documentation says that you can store an IPv4 address in a `INET6` column but need to convert it to an IPv6 address first, so we can't just change the `ipAddress()` type to `INET6`.

**Reply by juanparati** [reply]
@staudenmeir : Thank you for all the improvements!

To use native UUID with MariaDB it's a game changer, specially when partitions are used.
